Add foreground blurring (Privacy and Focus modes)

dshoreman opened the pull request:
Continuation/revival of #490 by @rajansaini691; see that PR for more background.

### Differences in this PR:
* Dropped merge commit, squashed formatting fixes
* Rebased onto latest `next` branch, resolved conflicts
* Added `--inactive-blur-include` option (see below)
* Added `--inactive-blur-opacity`
* Utilised the `blur_foreground` property (added but unused in original PR)
* Avoids blurring background *and* foreground of the same window to boost speed

### Option Rationale
Comments in the original PR called for an *exclusion* list to be added, but I couldn't see how that would be useful so opted for the inverse.  
Doing it this way allows for two different behaviours - either "privacy" or "focus" mode.

1. **Focus Mode**
  The easiest option, this will blur *all* inactive windows:
    ```ini
    inactive-blur = true;
    ```

2. **Privacy Mode**
  Combining both options will *only* blur windows defined in `--inactive-blur-list`.  
  This can be useful for hiding sensitive data, eg communication or banking apps:
    ```ini
    inactive-blur = true;
    inactive-blur-include = [
        "class_g = 'discord'",
        "name *= '- Gmail -'",
        "class_g = 'Signal'",
        "class_g = 'TelegramDesktop'",
    ];
    ```

Closes #489

Commit:
Add foreground blurring

Commit:
Add an inactive-blur option

Commit:
Check for inactive_blur when drawing fg blur

dshoreman committed:
Add --inactive-blur-include option

Adds an option to specify which windows will be blurred when inactive.

dshoreman committed:
Enable blurring all windows without a list

If --inactive-blur-include isn't specified, but --inactive-blur *is*,
then the sensible default would be to blur all inactive windows.

This commit does exactly that.

dshoreman committed:
Use `blur_foreground` to simplify conditionals

The property was already added in the first commit on this branch, but
it wasn't defined so it would've had no effect. With this, the long
conditional is easier to read and *should* update automatically.

dshoreman committed:
Avoid blurring background and foreground together

dshoreman committed:
Fix active window also being blurred

dshoreman commented:
Looking around I found two places that deal with the background blur: `win_set_blur_background` (called via ...determine.. in `win_on_factor_change`) and `win_blur_background` that actually draws it.

I *think* I've put enough checks for `w->blur_foreground` so it never does *both* blurs—even if the determine/set methods ever get written out—but I'd appreciate a sanity check here.

yshui commented on src/win.c:
```diff
@@ -1128,6 +1131,9 @@ void win_on_factor_change(session_t *ps, struct managed_win *w) {
 	// Focus needs to be updated first, as other rules might depend on the focused
 	// state of the window
 	win_update_focused(ps, w);
+	w->blur_foreground =
```
Need to add damage when blur_foreground changes

dshoreman replied:
I *think* 4bccb48 (based on `win_{determine,set}_blur_background()`) should solve this

yshui commented on src/backend/backend.c:
```diff
@@ -406,6 +406,18 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
+			// Blur window
+			if (w->blur_foreground) {
+				assert(ps->o.blur_method != BLUR_METHOD_NONE);
+				// FIXME Think more about combining blur w/ opacity
+				// FIXME Don't hardcode opacity
+				//        - needs to be set when focus_mode set to true
```
I'd like to see these resolved before merging this.

dshoreman replied:
Added an `--inactive-blur-opacity` option for this in 45dc995

yshui commented on src/config_libconfig.c:
```diff
@@ -566,6 +566,13 @@ char *parse_config_libconfig(options_t *opt, const char *config_file, bool *shad
+	// --inactive-blur
+	if (lcfg_lookup_bool(&cfg, "inactive-blur", &opt->inactive_blur)) {
+		if (opt->blur_method == BLUR_METHOD_NONE) {
+			opt->blur_method = BLUR_METHOD_BOX;
```
Same thing should be done in `options.c` too. Probably better to do this in one place after all config is parsed. Also, changing blue_method is not the right thing to do, what if the user doesn't want background blur?

dshoreman replied:
Moved the blur fallback code to picom.c in c586170, which seems to be the best place for it and fixes the segfault when using only CLI args.

Instead of overriding `blur_method`, it now uses a separate blur context. If `blur_method` is set then it copies `backend_blur_context`, otherwise a new context is created with the same box default as before.

yshui reviewed (changes requested):
Thanks for the PR! This is great work, I like how small the changes are for what it does.

Just a couple of concerns

dshoreman commented:
> Thanks for the PR! This is great work, I like how small the changes are for what it does.

Thanks! I can't take all the credit though. Much of the work is by @rajansaini691—I simply rebased and made some minor behavioural changes ;)

Sorry it took so long to fix the style issue, a few things came up all at once IRL. I'll see what I can do to resolve any issues later.  
It'd be nice to add a test or two as well but so far I haven't even *looked* inside the `tests` dir, so we'll see

obsidianical commented:
Hi, looking for an update on this, does this PR have a future?  I'm looking for this feature for quite a while and it's kind of disheartening to see this PR and its predecessor be practically dead.

rajansaini691 commented:
Hi Gabriel, I stopped work on this PR since I temporarily stopped using
Linux altogether for unrelated reasons. However, like my successor said, it
shouldn't be too much work to integrate my changes. Hope someone can pick
it up again and take focus mode to production! :)

On Mon, Sep 19, 2022, 4:40 AM Gabriel ***@***.***> wrote:

> Hi, looking for an update on this, does this PR have a future? I'm looking
> for this feature for quite a while and it's kind of disheartening to see
> this PR and its predecessor be practically dead.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yshui/picom/pull/765#issuecomment-1250906749>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJJ4AHNET5XQK3L3UIJ6KYTV7BGMBANCNFSM5MTRBCUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

obsidianical commented:
Hello, thanks for the response!
I'm both lacking the time and C skills to implement it sadly, but I use my laptop in public a lot so I hope that someone will integrate them soon.

dshoreman committed:
Patch for `mask` parameter in `blur` function

dshoreman commented:
@obsidianical Apologies, things came up after my first rebase in July and I forgot to push. I rebased again yesterday (only one new commit) and pushed just now after making sure it still works.

Most conflicts were from new `window-shadow-fg` and `window-shadow-fg-rule` options added in 654772b8cf956b645d3202b3671485714dd18f38. I'm unsure if it adds similar (or underlying) functionality that could replace or support this PR. It could be something else entirely...

---

I plan to throw a test config together later today after some sleep to try the plain `next` branch. If the functionality of this PR isn't already added by 654772b8cf956b645d3202b3671485714dd18f38 then I have no problem getting back to work resolving some of the original review issues.

dshoreman committed:
Add damage when blur_foreground changes

Moves the setting of `blur_foreground` to a new method which then calls
a second to actually set the value and apply damage if it's different,
similar to how the blur_background value/damage is handled.

dshoreman committed:
Fix segfault, don't hijack background blur

Rather than overriding the background blur method, this will only use
blur_method when it's already defined, otherwise creating a whole new
blur context using the 'box' blur method with default size of 3. This
fixes an issue where background blur would be forced even if unwanted.

By placing the fgcontext code in picom.c rather than config_libconfig.c,
this also fixes a segfault when passing `--inactive-blur` on the command
line without having a blur method set either in an option or in config.

dshoreman committed:
Add option to tweak inactive window blur opacity

dshoreman commented:
![image](https://user-images.githubusercontent.com/1521802/232639232-2c5b80a6-5535-49a4-96e0-4093bc5ca2a1.png)

Finally(!) found enough time to get back to work on this again :tada: 

@yshui I'm pretty sure everything's done now, but would you mind reviewing again in case I've missed anything? Thanks!
